labelbox/data/annotation_types/data/tiled_image.py

```diff
@@ -0,0 +1,525 @@
+        data.raise_for_status()
+        decoded = np.array(Image.open(BytesIO(data.content)))[..., :3]
+        if decoded.shape[:2] != (self.tile_size, self.tile_size):
+            logger.warning(f"Unexpected tile size {decoded.shape}.")
+        return decoded
+
+    def _crop_to_bounds(
+        self,
+        image: np.ndarray,
+        x_px_start: int,
+        y_px_start: int,
+        x_px_end: int,
+        y_px_end: int,
+    ) -> np.ndarray:
+        """This function slices off the excess pixels that are outside of the bounds.
+        This occurs because only full tiles can be downloaded at a time.
+        """
+
+        def invert_point(pt):
+            # Must have at least 1 pixel for stability.
+            pt = max(pt, 1)
+            # All pixel points are relative to a single tile
+            # So subtracting the tile size inverts the axis
+            pt = pt - self.tile_size
+            return pt if pt != 0 else None
+
+        x_px_end, y_px_end = invert_point(x_px_end), invert_point(y_px_end)
+        return image[y_px_start:y_px_end, x_px_start:x_px_end, :]
+
+    def _validate_num_tiles(self, xstart: float, ystart: float, xend: float,
+                            yend: float, max_tiles: int):
+        """Calculates the number of expected tiles we would fetch.
+
+        If this is greater than the number of max tiles, raise an error.
+        """
+        total_n_tiles = (yend - ystart + 1) * (xend - xstart + 1)
+        if total_n_tiles > max_tiles:
+            raise ValueError(f"Requested zoom results in {total_n_tiles} tiles."
+                             f"Max allowed tiles are {max_tiles}")
+
+    @validator('zoom_levels')
+    def validate_zoom_levels(cls, zoom_levels):
+        if zoom_levels[0] > zoom_levels[1]:
+            raise ValueError(
+                f"Order of zoom levels should be min, max. Received {zoom_levels}"
+            )
+        return zoom_levels
+
+
+class EPSGTransformer(BaseModel):
+    """Transformer class between different EPSG's. Useful when wanting to project
+    in different formats.
+
+    Requires as input a Point object.
+    """
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    transformer: Any
+
+    @staticmethod
+    def _is_simple(epsg: EPSG) -> bool:
+        return epsg == EPSG.SIMPLEPIXEL
+
+    @staticmethod
+    def _get_ranges(bounds: np.ndarray):
+        """helper function to get the range between bounds.
+        
+        returns a tuple (x_range, y_range)"""
+        x_range = np.max(bounds[:, 0]) - np.min(bounds[:, 0])
+        y_range = np.max(bounds[:, 1]) - np.min(bounds[:, 1])
+        return (x_range, y_range)
+
+    @staticmethod
+    def _min_max_x_y(bounds: np.ndarray):
+        """returns the min x, max x, min y, max y of a numpy array
+        """
+        return np.min(bounds[:, 0]), np.max(bounds[:, 0]), np.min(
+            bounds[:, 1]), np.max(bounds[:, 1])
```

Comment:
make sure to add typing hinting where necessary. These need return types and I noticed a few other missing hints.